// @react-awesome-query-builder/ui/cjs/components/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getWidgetId = exports.getRenderFromConfig = exports.dummyFn = exports.WithConfirmFn = exports.DragIcon = exports.Col = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _excluded = ["children"];
var Col = exports.Col = function Col(_ref) {
  var children = _ref.children,
    props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  return /*#__PURE__*/_react["default"].createElement("div", props, children);
};
var dummyFn = exports.dummyFn = function dummyFn() {};
var DragIcon = exports.DragIcon = function DragIcon() {
  return /*#__PURE__*/_react["default"].createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 24 24",
    fill: "gray",
    width: "18px",
    height: "18px"
  }, /*#__PURE__*/_react["default"].createElement("path", {
    d: "M0 0h24v24H0V0z",
    fill: "none"
  }), /*#__PURE__*/_react["default"].createElement("path", {
    d: "M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
  }));
};
var WithConfirmFn = exports.WithConfirmFn = function WithConfirmFn(Cmp) {
  return function (props) {
    var useConfirm = props.config.settings.useConfirm;
    var confirmFn = useConfirm ? useConfirm.call(props.config.ctx, props.config.ctx) : null;
    return /*#__PURE__*/_react["default"].createElement(Cmp, (0, _extends2["default"])({}, props, {
      confirmFn: confirmFn
    }));
  };
};
var getWidgetId = exports.getWidgetId = function getWidgetId(_ref2) {
  var id = _ref2.id,
    isLHS = _ref2.isLHS,
    delta = _ref2.delta,
    parentFuncs = _ref2.parentFuncs;
  return [id, isLHS ? "L" : "R", isLHS ? -1 : delta || 0, (parentFuncs || []).map(function (_ref3) {
    var _ref4 = (0, _slicedToArray2["default"])(_ref3, 2),
      f = _ref4[0],
      a = _ref4[1];
    return "".concat(f, "(").concat(a, ")");
  }).join("/")].join(":");
};
var getRenderFromConfig = exports.getRenderFromConfig = function getRenderFromConfig(config, renderFn) {
  var Cmp;
  if (typeof renderFn === "function") {
    Cmp = function Cmp(pr) {
      return renderFn === null || renderFn === void 0 ? void 0 : renderFn(pr, config.ctx);
    };
    Cmp.displayName = renderFn.name;
  } else {
    Cmp = renderFn;
  }
  return Cmp;
};
