// vimodev/dist/lib/Sketch/components/QueryBuilder.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireWildcard(require("react"));
var _GlobalContext = require("../contexts/GlobalContext");
var _mui = require("@react-awesome-query-builder/mui");
require("@react-awesome-query-builder/mui/css/styles.css");
require("./QueryBuilder.css");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; } // import "react-awesome-query-builder/lib/css/compact_styles.css";
var InitialConfig = _mui.MuiConfig;
delete InitialConfig["conjunctions"]["OR"];
InitialConfig["settings"]["showNot"] = false;
InitialConfig["settings"]["groupOperators"] = false;
InitialConfig["settings"]["canAddGroup"] = false;
InitialConfig["settings"]["canReorder"] = false;
InitialConfig["settings"]["renderSize"] = "small";
InitialConfig["settings"]["setOpOnChangeField"] = ["keep", "first"];
function QueryBuilder(_ref) {
  var _context$selectedSket, _context$selectedSket2;
  var NodeFields = _ref.NodeFields,
    EdgeFields = _ref.EdgeFields;
  var _useState = (0, _react.useState)(),
    _useState2 = _slicedToArray(_useState, 2),
    tree = _useState2[0],
    setTree = _useState2[1];
  var context = (0, _react.useContext)(_GlobalContext.AppContext);
  (0, _react.useEffect)(function () {
    var queryValue = {
      id: _mui.Utils.uuid(),
      type: "group"
    };
    setTree(_mui.Utils.checkTree(_mui.Utils.loadTree(queryValue), _objectSpread(_objectSpread({}, InitialConfig), {}, {
      fields: NodeFields
    })));
  }, []);
  (0, _react.useEffect)(function () {
    var queryValue = {
      id: _mui.Utils.uuid(),
      type: "group"
    };
    if (context.selectedSketchElement && context.selectedSketchElement.type === "edge") {
      setTree(_mui.Utils.checkTree(_mui.Utils.loadTree(queryValue), _objectSpread(_objectSpread({}, InitialConfig), {}, {
        fields: EdgeFields
      })));
    } else {
      setTree(_mui.Utils.checkTree(_mui.Utils.loadTree(queryValue), _objectSpread(_objectSpread({}, InitialConfig), {}, {
        fields: NodeFields
      })));
    }
  }, [context.selectedSketchElement]);
  var renderBuilder = function renderBuilder(props) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "query-builder-container",
      style: {
        padding: 0,
        minWidth: 300
      }
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "query-builder qb-lite",
      style: {
        margin: 0
      }
    }, /*#__PURE__*/_react["default"].createElement(_mui.Builder, props)));
  };
  var onChange = function onChange(immutableTree, config) {
    // Tip: for better performance you can apply `throttle` - see `examples/demo`
    // this.setState({, config: config});
    var query = _mui.Utils.mongodbFormat(immutableTree, config);
    setTree(immutableTree);
    var updatedElem = _objectSpread(_objectSpread({}, context.selectedSketchElement), {}, {
      tree: immutableTree,
      properties: query
    });
    context.setSelectedSketchElement(updatedElem);
  };
  return /*#__PURE__*/_react["default"].createElement("div", null, tree && (context === null || context === void 0 || (_context$selectedSket = context.selectedSketchElement) === null || _context$selectedSket === void 0 ? void 0 : _context$selectedSket.type) === "node" && (Object.keys(NodeFields).length > 0 ? /*#__PURE__*/_react["default"].createElement(_mui.Query, _extends({}, InitialConfig, {
    fields: NodeFields,
    value: context.selectedSketchElement.tree || tree,
    onChange: onChange,
    renderBuilder: renderBuilder
  })) : /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      height: "45px",
      width: "290px"
    }
  })), tree && (context === null || context === void 0 || (_context$selectedSket2 = context.selectedSketchElement) === null || _context$selectedSket2 === void 0 ? void 0 : _context$selectedSket2.type) === "edge" && (Object.keys(EdgeFields).length > 0 ? /*#__PURE__*/_react["default"].createElement(_mui.Query, _extends({}, InitialConfig, {
    fields: EdgeFields,
    value: context.selectedSketchElement.tree || tree,
    onChange: onChange,
    renderBuilder: renderBuilder
  })) : /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      height: "45px",
      width: "290px"
    }
  })));
}
var _default = exports["default"] = QueryBuilder;
