// @react-awesome-query-builder/mui/cjs/widgets/core/MuiFieldAutocomplete.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _MuiAutocomplete = _interopRequireDefault(require("../value/MuiAutocomplete"));
var _excluded = ["items", "selectedKey", "setField", "isValueField", "selectedLabel", "selectedOpts", "selectedAltLabel", "selectedFullLabel"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// see type FieldItemSearchableKeys
var mapFieldItemToOptionKeys = {
  key: "_value2",
  path: "value",
  label: "title",
  altLabel: "_altLabel",
  tooltip: "tooltip",
  grouplabel: "groupTitle"
};
var itemsToListValues = function itemsToListValues(items) {
  var level = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return items.map(function (item) {
    var items = item.items,
      path = item.path,
      key = item.key,
      label = item.label,
      altLabel = item.altLabel,
      disabled = item.disabled,
      grouplabel = item.grouplabel,
      group = item.group,
      matchesType = item.matchesType,
      tooltip = item.tooltip;
    var prefix = "\xA0\xA0".repeat(level);
    if (items) {
      return itemsToListValues(items, level + 1);
    } else {
      return {
        title: label,
        renderTitle: matchesType ? /*#__PURE__*/_react["default"].createElement("b", null, prefix + label) : prefix + label,
        value: path,
        disabled: disabled,
        groupTitle: level > 0 ? prefix + grouplabel : null,
        group: level > 0 ? _objectSpread(_objectSpread({}, group), {}, {
          label: prefix + group.label
        }) : null,
        tooltip: tooltip,
        _value2: key,
        _altLabel: altLabel
      };
    }
  }).flat(Infinity);
};
var fieldAdapter = function fieldAdapter(_ref, config) {
  var items = _ref.items,
    selectedKey = _ref.selectedKey,
    setField = _ref.setField,
    isValueField = _ref.isValueField,
    selectedLabel = _ref.selectedLabel,
    selectedOpts = _ref.selectedOpts,
    selectedAltLabel = _ref.selectedAltLabel,
    selectedFullLabel = _ref.selectedFullLabel,
    rest = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  var tooltipText = selectedAltLabel || selectedFullLabel;
  if (tooltipText == selectedLabel) tooltipText = null;
  var listValues = itemsToListValues(items);
  var value = selectedKey;
  var setValue = function setValue(value, _asyncValues) {
    if (!value && !isValueField) return undefined;
    return setField(value);
  };
  var filterOptionsConfig = {
    stringify: (0, _react.useCallback)(function (option) {
      var keysForFilter = config.settings.fieldItemKeysForSearch.map(function (k) {
        return mapFieldItemToOptionKeys[k];
      });
      var valueForFilter = keysForFilter.map(function (k) {
        return typeof option[k] == "string" ? option[k] : "";
      }).join("\0");
      return valueForFilter;
    }, [config])
  };
  return _objectSpread(_objectSpread({}, rest), {}, {
    tooltipText: tooltipText,
    listValues: listValues,
    setValue: setValue,
    filterOptionsConfig: filterOptionsConfig,
    allowCustomValues: false,
    multiple: false,
    disableClearable: !isValueField,
    value: value,
    isFieldAutocomplete: true
  });
};
var _default = exports["default"] = function _default(props) {
  return /*#__PURE__*/_react["default"].createElement(_MuiAutocomplete["default"], fieldAdapter(props, props.config));
};
