// @react-awesome-query-builder/core/cjs/import/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "_loadFromJsonLogic", {
  enumerable: true,
  get: function get() {
    return _jsonLogic._loadFromJsonLogic;
  }
});
Object.defineProperty(exports, "_loadFromSpel", {
  enumerable: true,
  get: function get() {
    return _spel._loadFromSpel;
  }
});
Object.defineProperty(exports, "getTree", {
  enumerable: true,
  get: function get() {
    return _tree.getTree;
  }
});
Object.defineProperty(exports, "isImmutableTree", {
  enumerable: true,
  get: function get() {
    return _tree.isImmutableTree;
  }
});
Object.defineProperty(exports, "isJsonLogic", {
  enumerable: true,
  get: function get() {
    return _tree.isJsonLogic;
  }
});
Object.defineProperty(exports, "isTree", {
  enumerable: true,
  get: function get() {
    return _tree.isTree;
  }
});
Object.defineProperty(exports, "jsToImmutable", {
  enumerable: true,
  get: function get() {
    return _tree.jsToImmutable;
  }
});
Object.defineProperty(exports, "loadFromJsonLogic", {
  enumerable: true,
  get: function get() {
    return _jsonLogic.loadFromJsonLogic;
  }
});
Object.defineProperty(exports, "loadFromSpel", {
  enumerable: true,
  get: function get() {
    return _spel.loadFromSpel;
  }
});
Object.defineProperty(exports, "loadTree", {
  enumerable: true,
  get: function get() {
    return _tree.loadTree;
  }
});
var _tree = require("./tree");
var _jsonLogic = require("./jsonLogic");
var _spel = require("./spel");
